src/uct/rocm/rocm_cma_ep.c

```diff
@@ -0,0 +1,297 @@
+                fn_name, (void *)remote_addr, (void*)key->address);
+
+    do {
+        iov_it_length = 0;
+        local_iov_it = 0;
+        for (iov_it = 0; iov_it < ucs_min(UCT_SM_MAX_IOV, iovcnt); ++iov_it) {
+            iov_slice_delivered = 0;
+
+            /* Get length of the particular iov element */
+            iov_slice_length = uct_iov_get_length(iov + iov_it);
+
+            /* Skip the iov element if no data */
+            if (!iov_slice_length) {
+                continue;
+            }
+            iov_it_length += iov_slice_length;
+
+            if (iov_it_length <= delivered) {
+                continue; /* Skip the iov element if transferred already */
+            } else {
+                /* Let's assume the iov element buffer can be delivered partially */
+                if ((iov_it_length - delivered) < iov_slice_length) {
+                    iov_slice_delivered = iov_slice_length - (iov_it_length - delivered);
+                }
+            }
+
+            local_ptr[local_iov_it]             = (void *)((char *)iov[iov_it].buffer +
+                                                        iov_slice_delivered);
+            local_iov[local_iov_it].SizeInBytes = iov_slice_length - iov_slice_delivered;
+
+            /* It is possible that we get host (CPU) address as local address.
+             * We need to get GPU address to be used for CMA operation.
+             * If this is memory was not yet registered with ROCm stack and
+             * flag "any_memory" is set than lock this memory.
+             */
+            ucs_status_t ucs_status = uct_rocm_cma_ptr_to_gpu_ptr(local_ptr[local_iov_it],
+                                            &local_iov[local_iov_it].MemoryAddress,
+                                            local_iov[local_iov_it].SizeInBytes,
+                                            rocm_md->any_memory,
+                                            &local_ptr_locked[local_iov_it]);
+
+            if (ucs_status != UCS_OK) {
+                uct_rocm_cma_unlock_ptrs(local_ptr, local_ptr_locked, local_iov_it);
+                return ucs_status;
+            }
+
+            ucs_trace("uct_rocm_cma_ep_common_zcopy: [%d] Local address %p (GPU ptr %p), Local Size 0x%x",
+                                (int) local_iov_it,
+                                local_ptr[local_iov_it],
+                                local_iov[local_iov_it].MemoryAddress,
+                                (int) local_iov[local_iov_it].SizeInBytes);
+
+            ++local_iov_it;
+        }
+
+        if (!delivered) {
+            length = iov_it_length; /* Keep total length of the iov buffers */
+        }
+
+        if(!length) {
+            return UCS_OK;          /* Nothing to deliver */
+        }
+
+        remote_iov.MemoryAddress = (void *)(key->address + delivered);
+        remote_iov.SizeInBytes   = length - delivered;
+
+        HSAKMT_STATUS hsa_status = fn_p(ep->remote_pid,
```

Comment:
It looks like a critical path function. Would it make a sense to avoid pointer to function (which prevents inlining) and go with inlining ? What is the overall overhead of this function call in respect to the operation.